### ISSUE TITLE
Masque errors

### ISSUE THREAD

**Issue by switchza** [cogwerkz/gUI4#40]
I'm unable to use Masque at all alongside gUI. With only "Goldpaw's UI Core" and "Masque" enabled, trying to load masque options throws this error:

Message: ...sque\Libs\AceGUI-3.0\widgets\AceGUIWidget-Button.lua:21: Usage: PlaySound(soundKitID, optional["SFX","Music","Ambience", "Dialog" or "Master"], [forceNoDuplicates, runFinishCallback])
Time: 12/17/17 14:42:23
Count: 1
Stack: ...sque\Libs\AceGUI-3.0\widgets\AceGUIWidget-Button.lua:21: Usage: PlaySound(soundKitID, optional["SFX","Music","Ambience", "Dialog" or "Master"], [forceNoDuplicates, runFinishCallback])
[C]: ?
[C]: ?
...sque\Libs\AceGUI-3.0\widgets\AceGUIWidget-Button.lua:21: in function <...sque\Libs\AceGUI-3.0\widgets\AceGUIWidget-Button.lua:19>

Locals:

**Comment by goldpaw** [comment]
Your version of Masque is using an outdated embedded version of the Ace3 libraries. Blizzard did some changes to the PlaySound commands in patch 7.3.0, and a lot of addon developers still haven't updated their libraries to work around this. 

Good news is you can fix it by installing the standalone Ace3 addon and restarting WoW:
https://www.curseforge.com/wow/addons/ace3

The updated version should overwrite the older embedded one.

**Comment by switchza** [comment]
Thanks for the response! I've installed the standalone Lib: Ace3 (from curseforge 1 Oct 2017) and restarted wow completely - and I'm still getting the same error. :(

I've also just tried copying and overwriting the Ace* folders under Masque\Libs without luck - same error. I guess this probably has nothing to do with your UI though. I just noticed it only happens when gUI is enabled.

**Comment by switchza** [comment]
Update: I'm getting the same errors with NPCScan (when changing tabs in addon options), also only when gUI Core is enabled. 

I did some investigation into the AceGUIWidget-Button.lua present in \libs\AceGUI for both of these addons, and they are both latest (matching latest release of Ace3). Disabling gUI results in both Addons working normally.

My LUA and addon framework knowledge is limited, so this could be the wrong way round.. but could it be that gUI is changing the underlying/shared Ace3 version which is used by these addons, causing the error? I'm only thinking that because they both immediately stop throwing errors when gUI is disabled and /reload.

**Comment by goldpaw** [comment]
Figured it out. 

The bug is a result of how Ace3 checks for WoW client version here. They don't. They base what they put into the PlaySound command on whether or not the command PlaySoundKitID exists. That was removed in patch 7.3.0, but Goldpaw's UI adds it back and thus Ace3 assumes it's still on an older patch, and calls the PlaySound command with wrong arguments. 

Could have been avoided if Ace3 simply had done a client version check like we're supposed to, instead of doing silly guesswork and assumptions. But they didn't, so I guess it's down to me to write a different fix for Goldpaw's UI that works with the Ace3 way. 

I'm on it! :)

**Comment by switchza** [comment]
You're the best! 

For others, I found an addon called PlaySoundFixer (https://www.curseforge.com/wow/addons/playsoundfixer) which allows me to at least keep using the other addons along  with gUI in the interim.

**Comment by goldpaw** [comment]
Do NOT use addons that "fixes" playsound! They replace the API call, which cause more bugs than it fix! Pretty much any actionbar, unitframe or any secure code that tries to make a sound will become tainted and break. Tainted more or less means "changed by a non-Blizzard addon", which makes Blizzard thinks something has been tinkered with, and thus anything that was trying to run something tainted will simply stop running. 

One example is logging out. In many cases you won't be able to click the "log out" button on the main menu when PlaySound is tainted. My advice is to keep the updated Ace3 standalone around, and simply drop any addon that hasn't been updated to work properly yet. Or the headaches will never stop. ;)

**Comment by switchza** [comment]
Ok noted! The addons themselves are updated though, it sounds like Ace3 itself is the issue here, at least when used alongside gUI. So standalone Ace3 doesn't help me.

I'm just using Goldpaw UI + the recommended list in your Curseforge page (Masque, Minimapbutton, etc) and NPCScan.

What kind of mission is a fix like this for you? Is it something that can be put together as a patch (I'd even be prepared to help out), or is it a major refactor.

**Comment by goldpaw** [comment]
We're talking about the fix to the issue you reported in this thread? I've already fixed it, the most recent version of Goldpaw's UI (1.1.33) here on GitHub includes it. It'll be live on CurseForge within a few hours at most, it's currently going through their approval process. :) 

I could use both Goldpaw's UI, Masque and open the options without problems just now, without any bugs or conflicts: 

![screenshot from 2017-12-19 12-12-56](https://user-images.githubusercontent.com/6476535/34158504-011a6982-e4c6-11e7-9022-4e04863af109.png)

**Comment by switchza** [comment]
Ok I got confused, I see now you were warning about using playsound fix addons in general. I'll grab the latest GitHub version in a couple hours to test! Thank you so much Goldpaw!

**Comment by goldpaw** [comment]
Yeah, the bugs you reported here weren't actually PlaySound bugs, they were a result of me and Ace3 adjusting to wow patch 7.3.0 in different and incompatible ways. I fixed that, and the fix will be live on Curse/Twitch too soon. 

When I first replied in this thread, I thought we were dealing with some addons that hadn't yet updated their Ace3 libraries to patch 7.3.0, which is usually the case with PlaySound-related bugs. But of course I eventually realized it was my own addon that was incompatible with the updated Ace3. Bummer. But it's fixed now! :D 

When it comes to so called PlaySound-"fixers", that's another issue. They were meant as a very short term band aid for addons that couldn't be fixed by updating Ace3. The problem is that they all work by replacing the Blizzard PlaySound command, which is a big no-no, as it will cause any "secure" action that tries to play a sound to break. Secure actions are anything that usually require user input, like clicking actionbuttons, changing your target, logging out, etc. And I don't know about you, but I'd rather stop using an addon that stop using my unitframes, actionbuttons and the ability to log out, lol! :D 

So to put it short, apart from the incompability between gUI4 and Ace3 which I've fixed now, there are only two ways to go with PlaySound-bugs: See if the updated Ace3 can fix it, or accept that the addon is broken in patch 7.3.0 and stop using it.
